File: app/src/main/java/com/sleticalboy/dailywork/ui/activity/FaceActivity.java
package com.sleticalboy.dailywork.ui.activity;

import android.app.Activity;
import android.hardware.Camera;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.RelativeLayout;

import com.sleticalboy.dailywork.R;
import com.sleticalboy.dailywork.base.BaseActivity;
import com.sleticalboy.dailywork.util.CameraUtils;
import com.sleticalboy.dailywork.util.FaceDetectionListenerImpl;
import com.sleticalboy.dailywork.weight.FaceView;
import com.sleticalboy.dailywork.weight.VerificationSurfaceView;

import java.lang.ref.WeakReference;

/**
 * Created on 18-2-24.
 *
 * @author sleticalboy
 * @version 1.0
 * @description
 */
public class FaceActivity extends BaseActivity {

    ImageView btnSwitchCamera;
    ImageView btnSwitchFlash;
    VerificationSurfaceView verifySurfaceView;
    FaceView mFaceView;
    ImageButton btnTakePic;
    FrameLayout flCameraFocus;
    RelativeLayout rlSettings;

    private Handler mHandler = new MainHandler(this);
    private int mHeight;
    private int mWidth;

    @Override
    protected void initData() {
    }

    @Override
    protected void initView() {
        rlSettings = findViewById(R.id.rlSettings);
        btnSwitchCamera = findViewById(R.id.btnSwitchCamera);
        btnSwitchFlash = findViewById(R.id.btnSwitchFlash);

        verifySurfaceView = findViewById(R.id.verifySurfaceView);
        verifySurfaceView.setHandler(mHandler);

        mFaceView = findViewById(R.id.pictureView);
        btnSwitchCamera = findViewById(R.id.btnSwitchCamera);
        btnTakePic = findViewById(R.id.btnTakePic);

        flCameraFocus = findViewById(R.id.flCameraFocus);
        // width 和 height 是具体数值时可以用此种方法测量
        int widthMeasureSpec = View.MeasureSpec.makeMeasureSpec(120, View.MeasureSpec.EXACTLY);
        int heightMeasureSpec = View.MeasureSpec.makeMeasureSpec(120, View.MeasureSpec.EXACTLY);
        flCameraFocus.measure(widthMeasureSpec, heightMeasureSpec);
        mWidth = flCameraFocus.getMeasuredWidth();
        mHeight = flCameraFocus.getMeasuredHeight();
        Log.d("measure", "mWidth:" + mWidth);
        Log.d("measure", "mHeight:" + mHeight);

        flCameraFocus.getViewTreeObserver().addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
            @Override
            public boolean onPreDraw() {
                // visibility 属性是 invisibility 或者 gone 时测量结果为 0
                mWidth = flCameraFocus.getMeasuredWidth() / 2;
                mHeight = flCameraFocus.getMeasuredHeight() / 2;
                Log.d("onPreDraw", "mWidth:" + mWidth);
                Log.d("onPreDraw", "mHeight:" + mHeight);
                flCameraFocus.getViewTreeObserver().removeOnPreDrawListener(this);
                return true;
            }
        });

        initListeners();
    }

    // 当 Activity 的窗口得到或者失去焦点时均会被调用一次，即当 Activity 继续执行和暂停执行时，此方法会被调用
    // 也就是说当频繁的调用 onResume 和 onPause 方法时，此方法会被频繁调用
    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            // visibility 属性是 invisibility 或者 gone 时测量结果为 0
            mWidth = flCameraFocus.getMeasuredWidth() / 2;
            mHeight = flCameraFocus.getMeasuredHeight() / 2;
            Log.d("onWindowFocusChanged", "mWidth:" + mWidth);
            Log.d("onWindowFocusChanged", "mHeight:" + mHeight);
        }
    }

    private void initListeners() {
        btnTakePic.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                CameraUtils.getInstance().takePicture();
            }
        });
        btnSwitchCamera.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                changeCamera();
            }
        });
        btnSwitchFlash.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                CameraUtils.getInstance().setFlashMode((ImageView) v);
            }
        });

        verifySurfaceView.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                if (CameraUtils.getInstance().getCameraInfo().facing == Camera.CameraInfo.CAMERA_FACING_BACK) {
                    return mGestureDetector.onTouchEvent(event);
                } else {
                    return false;
                }
            }
        });
    }

    private void changeCamera() {
        CameraUtils.getInstance().stopPreview();
        int newCameraId = (CameraUtils.getInstance().getCameraId() + 1) % 2;
        CameraUtils.getInstance().openCamera(newCameraId);
        CameraUtils.getInstance().startPreview(verifySurfaceView.getHolder());
        if (newCameraId == Camera.CameraInfo.CAMERA_FACING_BACK) {
            btnSwitchCamera.setImageResource(R.mipmap.camera_setting_switch_back);
            btnSwitchFlash.setVisibility(View.VISIBLE);
        } else {
            btnSwitchCamera.setImageResource(R.mipmap.camera_setting_switch_front);
            btnSwitchFlash.setVisibility(View.GONE);
        }
    }

    private void startDetect() {
        Camera.Parameters params = CameraUtils.getInstance().getCameraParameters();
        Camera camera = CameraUtils.getInstance().getCamera();
        if (params.getMaxNumDetectedFaces() > 0) {
            if (mFaceView != null) {
                mFaceView.clearFaces();
                mFaceView.setVisibility(View.GONE);
            }
        }
        camera.setFaceDetectionListener(new FaceDetectionListenerImpl(mHandler));
        camera.startFaceDetection();
    }

    @Override
    protected int attachLayout() {
        return R.layout.activity_face;
    }

    static class MainHandler extends Handler {

        private WeakReference<Activity> mActivity;

        MainHandler(Activity activity) {
            mActivity = new WeakReference<>(activity);
        }

        @Override
        public void handleMessage(Message msg) {
            final FaceActivity activity = (FaceActivity) mActivity.get();
            switch (msg.what) {
                case CameraUtils.MSG_PREVIEW_STARTED:
                    activity.startDetect();
                    break;
                case CameraUtils.MSG_SCAN_FACE:
                    activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            Camera.Face[] faces = (Camera.Face[]) msg.obj;
                            activity.mFaceView.setFaces(faces);
                        }
                    });
                    break;
            }
        }
    }

    private GestureDetector mGestureDetector = new GestureDetector(new GestureDetector.SimpleOnGestureListener() {
        @Override
        public boolean onSingleTapUp(MotionEvent event) {
            CameraUtils.getInstance().autoFocus(new Camera.AutoFocusCallback() {
                @Override
                public void onAutoFocus(boolean success, Camera camera) {
                    if (success) {
                        Log.e("FaceActivity", "聚焦成功");
                    } else {
                        Log.e("FaceActivity", "聚焦失败");
                    }
                    flCameraFocus.setVisibility(View.GONE);
                }
            });
            CameraUtils.getInstance().setFocusArea(FaceActivity.this, event);
            showFocusLayout(event);
            return true;
        }
    });

    private void showFocusLayout(MotionEvent event) {
        int x = (int) event.getX();
        int y = (int) event.getY();
        FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) flCameraFocus.getLayoutParams();
        params.leftMargin = (int) (x - mWidth + 0.5);
        params.topMargin = (int) (y - mHeight + 0.5 + rlSettings.getHeight());
        flCameraFocus.requestLayout(); // flCameraFocus.setLayoutParams(params);
        flCameraFocus.setVisibility(View.VISIBLE);
    }
}

File: MainApp/src/main/java/com/sleticalboy/dailywork/ui/activity/FaceActivity.kt
package com.sleticalboy.dailywork.ui.activity

import android.app.Activity
import android.hardware.Camera
import android.os.Handler
import android.os.Message
import android.util.Log
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.RelativeLayout

import com.sleticalboy.dailywork.R
import com.sleticalboy.dailywork.base.BaseActivity
import com.sleticalboy.dailywork.util.CameraUtils
import com.sleticalboy.dailywork.util.FaceDetectionListenerImpl
import com.sleticalboy.dailywork.weight.FaceView
import com.sleticalboy.dailywork.weight.VerificationSurfaceView

import java.lang.ref.WeakReference

/**
 * Created on 18-2-24.
 *
 * @author sleticalboy
 * @version 1.0
 * @description
 */
class FaceActivity : BaseActivity() {

    private var btnSwitchCamera: ImageView? = null
    private var btnSwitchFlash: ImageView? = null
    private var verifySurfaceView: VerificationSurfaceView? = null
    private var mFaceView: FaceView? = null
    private var btnTakePic: ImageButton? = null
    private var flCameraFocus: FrameLayout? = null
    private var rlSettings: RelativeLayout? = null

    private val mHandler = MainHandler(this)
    private var mHeight: Int = 0
    private var mWidth: Int = 0

    private val mGestureDetector = GestureDetector(object : GestureDetector.SimpleOnGestureListener() {

        override fun onDown(e: MotionEvent?): Boolean {
            Log.d("action", "down")
            return true
        }

        override fun onSingleTapUp(event: MotionEvent): Boolean {
            Log.d("touch", "single touch")
            CameraUtils.instance.autoFocus(Camera.AutoFocusCallback { success, _ ->
                if (success) {
                    Log.d("FaceActivity", "聚焦成功")
                } else {
                    Log.d("FaceActivity", "聚焦失败")
                }
                flCameraFocus!!.visibility = View.GONE
            })
            CameraUtils.instance.setFocusArea(this@FaceActivity, event)
            showFocusLayout(event)
            return true
        }
    })

    override fun initData() {}

    override fun initView() {
        rlSettings = findViewById<View>(R.id.rlSettings) as RelativeLayout
        btnSwitchCamera = findViewById<View>(R.id.btnSwitchCamera) as ImageView?
        btnSwitchFlash = findViewById<View>(R.id.btnSwitchFlash) as ImageView?

        verifySurfaceView = findViewById<View>(R.id.verifySurfaceView) as VerificationSurfaceView?
        verifySurfaceView!!.handler = mHandler

        mFaceView = findViewById<View>(R.id.pictureView) as FaceView?
        btnSwitchCamera = findViewById<View>(R.id.btnSwitchCamera) as ImageView?
        btnTakePic = findViewById<View>(R.id.btnTakePic) as ImageButton?

        flCameraFocus = findViewById<View>(R.id.flCameraFocus) as FrameLayout?
        // width 和 height 是具体数值时可以用此种方法测量
        val widthMeasureSpec = View.MeasureSpec.makeMeasureSpec((1 shl 30) - 1, View.MeasureSpec.AT_MOST)
        val heightMeasureSpec = View.MeasureSpec.makeMeasureSpec((1 shl 30) - 1, View.MeasureSpec.AT_MOST)
        flCameraFocus!!.measure(widthMeasureSpec, heightMeasureSpec)
        mWidth = flCameraFocus!!.measuredWidth / 2
        mHeight = flCameraFocus!!.measuredHeight / 2
        Log.d("measure", "mWidth:" + mWidth)
        Log.d("measure", "mHeight:" + mHeight)

        initListeners()
    }

    // 当 Activity 的窗口得到或者失去焦点时均会被调用一次，即当 Activity 继续执行和暂停执行时，此方法会被调用
    // 也就是说当频繁的调用 onResume 和 onPause 方法时，此方法会被频繁调用
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            // visibility 属性是 invisibility 或者 gone 时测量结果为 0
            val width = flCameraFocus!!.measuredWidth / 2
            val height = flCameraFocus!!.measuredHeight / 2
            Log.d("onWindowFocusChanged", "width:" + width)
            Log.d("onWindowFocusChanged", "height:" + height)
        }
    }

    private fun initListeners() {
        btnTakePic!!.setOnClickListener { CameraUtils.instance.takePicture() }
        btnSwitchCamera!!.setOnClickListener { changeCamera() }
        btnSwitchFlash!!.setOnClickListener { v -> CameraUtils.instance.setFlashMode(v as ImageView) }

        verifySurfaceView!!.setOnTouchListener { _, event ->
            if (CameraUtils.instance.cameraInfo.facing == Camera.CameraInfo.CAMERA_FACING_BACK) {
                mGestureDetector.onTouchEvent(event)
            } else {
                false
            }
        }
    }

    private fun changeCamera() {
        CameraUtils.instance.stopPreview()
        val newCameraId = (CameraUtils.instance.cameraId + 1) % 2
        CameraUtils.instance.openCamera(newCameraId)
        CameraUtils.instance.startPreview(verifySurfaceView!!.holder)
        if (newCameraId == Camera.CameraInfo.CAMERA_FACING_BACK) {
            btnSwitchCamera!!.setImageResource(R.mipmap.camera_setting_switch_back)
            btnSwitchFlash!!.visibility = View.VISIBLE
        } else {
            btnSwitchCamera!!.setImageResource(R.mipmap.camera_setting_switch_front)
            btnSwitchFlash!!.visibility = View.GONE
        }
    }

    private fun startDetect() {
        val params = CameraUtils.instance.cameraParameters
        val camera = CameraUtils.instance.mCamera
        if (params!!.maxNumDetectedFaces > 0) {
            if (mFaceView != null) {
                mFaceView!!.clearFaces()
                mFaceView!!.visibility = View.GONE
            }
        }
        camera!!.setFaceDetectionListener(FaceDetectionListenerImpl(mHandler))
        camera.startFaceDetection()
    }

    override fun attachLayout(): Int {
        return R.layout.activity_face
    }

    internal class MainHandler(activity: Activity) : Handler() {

        private val mActivity: WeakReference<Activity> = WeakReference(activity)

        override fun handleMessage(msg: Message) {
            val activity = mActivity.get() as FaceActivity
            when (msg.what) {
                CameraUtils.MSG_PREVIEW_STARTED -> activity.startDetect()
                CameraUtils.MSG_SCAN_FACE -> activity.runOnUiThread {
                    val faces = msg.obj as Array<Camera.Face>
                    activity.mFaceView!!.setFaces(faces)
                }
            }
        }
    }

    // 点击屏幕显示聚焦图标
    private fun showFocusLayout(event: MotionEvent) {
        val x = event.x.toInt()
        val y = event.y.toInt()
        val params = flCameraFocus!!.layoutParams as FrameLayout.LayoutParams
        params.leftMargin = (x - mWidth + 0.5).toInt()
        Log.d("rlSettings'", "height = " + rlSettings!!.height)
        params.topMargin = ((y - mHeight).toDouble() + 0.5 /*+ rlSettings!!.height.toDouble()*/).toInt()
        flCameraFocus!!.layoutParams = params
        flCameraFocus!!.visibility = View.VISIBLE
    }
}
